receiver/datadogreceiver/internal/translator/sketches.go

```diff
@@ -0,0 +1,268 @@
+}
+
+// mapSketchBucketsToHistogramBuckets attempts to map the counts in each Sketch bucket to the closest equivalent Exponential Histogram
+// bucket(s). It works by first calculating an Exponential Histogram key that corresponds most closely with the Sketch key (using the lower
+// bound of the sketch bucket the key corresponds to), calculates differences in the range of the Sketch bucket and exponential histogram bucket,
+// and distributes the count to the corresponding bucket, and the bucket prior to it and after it, based on the proportion of overlap between the
+// exponential histogram buckets and the Sketch bucket. Note that the Sketch buckets are not separated into positive and negative buckets, but exponential
+// histograms store positive and negative buckets separately. Negative buckets in exponential histograms are mapped in the same way as positive buckets.
+// Note that negative indices in exponential histograms do not necessarily correspond to negative values; they correspond with values between 0 and 1,
+// on either the negative or positive side
+func mapSketchBucketsToHistogramBuckets(sketchKeys []int32, sketchCounts []uint32) (map[int]uint64, map[int]uint64, uint64) {
+	var zeroCount uint64
+
+	var positiveBuckets = make(map[int]uint64)
+	var negativeBuckets = make(map[int]uint64)
+
+	// The data format for the sketch received from the sketch payload does not have separate positive and negative buckets,
+	// and instead just uses a single list of sketch keys that are in order by increasing bucket index, starting with negative indices,
+	// which correspond to negative buckets
+	for i := range sketchKeys {
+		if sketchKeys[i] == 0 { // A sketch key of 0 corresponds to the zero bucket
+			zeroCount += uint64(sketchCounts[i])
+			continue
+		}
+
+		// The approach here is to use the Datadog sketch index's lower bucket boundary to find the
+		// OTel exponential histogram bucket that with the closest range to the sketch bucket. Then,
+		// the buckets before and after that bucket are also checked for overlap with the sketch bucket.
+		// A count proportional to the intersection of the sketch bucket with the OTel bucket(s) is then
+		// added to the OTel bucket(s). After looping through all possible buckets that are within the Sketch
+		// bucket range, the bucket with the highest proportion of overlap is given the remaining count
+		sketchLowerBound, sketchUpperBound := getSketchBounds(sketchKeys[i])
+		sketchBucketSize := sketchUpperBound - sketchLowerBound
+		histogramKey := sketchLowerBoundToHistogramIndex(sketchLowerBound)
+		highestCountProportion := 0.0
+		highestCountIdx := 0
+		targetBucketCount := uint64(sketchCounts[i])
+		var currentAssignedCount uint64
+
+		// In some cases, the exponential histogram index that is mapped from the Sketch index corresponds to a bucket that
+		// has a lower bound that is higher than the Sketch bucket's lower bound. In this case, it is necessary to start
```

Comment:
Is this possible? We find a histogram index that covers the lower bound with `sketchLowerBoundToHistogramIndex(sketchLowerBound)` so the histogram bucket's lower bound shouldn't ever be higher than the sketch's lower bound

Reply:
I think you're right Fiona. We're taking the Frexp of the sketch lower bound (B). There's two possibilities. Either the `frac` part is 0.5 , in which case it was a power of two and then we use that as the upper bound of the exponential bucket , the lower being lower obviously (for example if you calculate Frexp(2)-> (0.5, 2)-> we return 31, which is the (1.957, 2] bucket.

Otherwise we take the floor(log_2(B)*32) . When you calculate the bucket boundary: 2^(floor(log_2(B)*32)/32) it's going to give a lower number than 2^(log_2(B)*32/32) which is the original number.

Reply:
That's fair. I think the reason I implemented it to start in the bucket below the histogram index mapped from the sketch index was in case of floating point rounding errors. I can update it to start at the histogram index that is computed from the sketch index, though